azure-kusto-data/azure/kusto/data/_token_providers.py

```diff
@@ -39,22 +50,38 @@ class TokenProviderBase(abc.ABC):
     _kusto_uri = None
     _cloud_info = None
     _scopes = None
+    lock = Lock()
 
-    def __init__(self, kusto_uri: str):
+    def __init__(self, kusto_uri: str, is_async: bool = False):
+        self.is_async = is_async
         self._kusto_uri = kusto_uri
         if kusto_uri is not None:
             self._scopes = [kusto_uri + ".defult" if kusto_uri.endswith("/") else kusto_uri + "/.default"]
             self._cloud_info = CloudSettings.get_cloud_info()
 
     def get_token(self):
         """ Get a token silently from cache or authenticate if cached token is not found """
+
+        with TokenProviderBase.lock:

```

Comment:
Pleas use a double check lock pattern

Reply:
Like this?

Reply:
Yes

Reply:
The double-checked locking pattern is a useful pattern in some languages, but notoriously not thread-safe in others.

Coincidentally, our own Michael Spector suggested in https://stackoverflow.com/questions/55957708/is-double-checked-locking-thread-safe-in-python that the value of this pattern depends on the specific Python implementation.